### PR TITLE
 feat(sketch): add support for categories in symbols

### LINKED ISSUE #3406: Finalize Elements Plugin output
Based on the output of #3405, this issue is for resolving the discrepancies with naming conventions in code for all of the Sketch commands. The end result of this will be the final plugin that will be used by design to generate the Elements Kit.

- [ ] Sync subcategories for icons

### PR DESCRIPTION
**Pull request by joshblack**
Closes #3406

Add support for category names in symbols

#### Changelog

**New**

**Changed**

- Update our metadata scaffold script to include categories in variants
- Update icon plugin command to include categories in symbol name

**Removed**

### CODE REVIEW

**Comment by joshblack** [comment]
Note to reviewers:

- [x] I need to remove the TODO blocks
- [x] I need to update the Joi schema for the metadata
- [ ] ~Might be worth adding a macos step for testing sketch plugin stuff~

Also, this should be rebased and merged if you see this PR as green and look to merge it in 👍

**Review by dakahn — APPROVED** [review]

File: packages/sketch/src/commands/icons/generate.js [hunk 1]
```diff
@@ -188,8 +225,9 @@ function normalize(icons) {
     if (!icon.size) {
       return acc;
     }
+    const name = icon.basename;
     // Drop size from prefix
-    const name = [...icon.prefix.slice(1), icon.basename].join('/');
+    // const name = [...icon.prefix.slice(1), icon.basename].join('/');
```
Comment (joshblack) [inline packages/sketch/src/commands/icons/generate.js]: Need to remove

File: packages/sketch/src/commands/icons/generate.js [hunk 2]
```diff
@@ -96,16 +123,21 @@ export function generate() {
           },
         };
 
-        let symbolName =
-          sizes.length !== 1
-            ? `category/${name}/${icon.size}`
-            : `category/${name}`;
+        // TODO: remove optional object until we figure out namespace issue
```
Comment (joshblack) [inline packages/sketch/src/commands/icons/generate.js]: Need to remove and update `|| {}` below

**Review by vpicone — APPROVED** [review]

**Review by emyarod — COMMENTED** [review]
are the test failures related to Joi schema updates?

**Comment by joshblack** [comment]
@emyarod yeah, need to update the schema change I made to metadata!

**Review by emyarod — APPROVED** [review]
thanks for updating the schema, looks good to me

**Comment by joshblack** [comment]
Note to anyone looking to merge this in, still have the following TODOs:

- [x] Make sure type tokens are up-to-date
- [x] Separate out icon sync from icon page generation
- [x] Update colors page layout
- [x] Update page names to be lowercase
- [ ] ~Add group named `alert` for orange/yellow colors~ manual step for now
- [x] Add `theme` prefix to token names
- [x] Lowercase theme names
- [x] Remove dashes in swatch names
- [x] Edit type styles to use no border
- [ ] Get final sign-off from @laurenmrice about the generated commands

**Review by vpicone — DISMISSED** [review]
Getting the following warning for glyphs when sourcing the data through gatsby. I think we should either come up with an integer representation of glyphs or have the sizes be strings.

```
warn There are conflicting field types in your data.

If you have explicitly defined a type for those fields, you can safely ignore this warning message.
Otherwise, Gatsby will omit those fields from the GraphQL schema.

If you know all field types in advance, the best strategy is to explicitly define them with the `createTypes` action, and skip inference with the `@dontInfer`
 directive.
See https://www.gatsbyjs.org/docs/actions/#createTypes
DataJson.icons.variants.sizes:
 - type: [number,string]
   value: [ 16, ..., 'glyph' ]
   source: File "src/data/metadata.json"
 - type: [number]
   value: [ 16, ... ]
   source: File "src/data/metadata.json"
```

**Comment by joshblack** [comment]
@vpicone this seems like a Gatsby-specific issue right? What would we gain from changing the authorship format to adjust for a specific tool? Seems appropriate for numbers to be numbers and glyph to be a string.

**Comment by vpicone** [comment]
@joshblack It's just a warning, and not really anything to do with Gatsby (gatsby just provided the warning). Can definitely just provide gatsby with a custom type. In general, mixing types in an array seems like bad practice. Performing any operations on them becomes treacherous.

You can imagine someone accidentally using the sizes array and getting something like `width: 'glyphpx'`

**Comment by joshblack** [comment]
@vpicone I believe we'd still have this issue even with the suggested changes. For example, if we converted to all being a string we would have to conditionally check when parsing the value if it is indeed a number or not.

```js
const sizes = ['16', '32', 'glyph'];
// Oops
const numberedSizes = sizes.map(size => parseInt(size, 10));
// Maybe?
const translatedSizes = sizes.map(size => {
  if (isNaN(size)) {
    return size;
  }
  return parseInt(size, 10);
});

```

This is to say that no matter what we choose we will need to have a logic gate that checks if an icon is sized versus is a glyph.

**Comment by vpicone** [comment]
@joshblack true, I suppose it only becomes an issue for string/number specific functions.

I was under the impression a glyph meant it only existed at one size. Wouldn't that just mean it only has one size in the array? and maybe a `glyph: true` property? `glyph` has a lot more meaning than just the other numbers in the array

**Comment by joshblack** [comment]
@vpicone icons could have glyph variants in addition to their boxed sizes, for example chevron has `[16, 20, 24, 32, 'glyph']`.

**Comment by vpicone** [comment]
And we don't know what size that is? Are all glyphs the same size? Or does it just mean "some bespoke size we don't know about"

**Comment by joshblack** [comment]
@vpicone all glyph sizes are different and bespoke to the asset 👍 We could find it for a particular glyph, but they often are not in a "box" shape and so we would need to express width and height. The sizes option here is mostly for grouping things up by size (like in the existing website) and so understanding the exact dimensions has not been necessary under these requirements

**Comment by vpicone** [comment]
@joshblack is there anything wrong with making all the constituents of the size array the same type? If so, feel free to merge.

**Comment by joshblack** [comment]
@vpicone what I'm curious about is what benefits we get from this change. This code doesn't exist in isolation, so unfortunately things will need to be updated with any changes we make, but if the benefit is worth it then we should 100% update it. As I mentioned earlier, it seems like most implementations would still need to have a check regardless of if we changed the format.

**Comment by vpicone** [comment]
Just general code quality, when performing operations on an array I generally expect the types to be the same. Could just be personal preference but 'glyph' tells you a lot more than the sibling members of the array.

Since we're generating all sizes, the sizes array in the build artifact is superfluous anyway (they'd all just be [16,20,24,32]). We could just use `glyph:true` to indicate a glyph of bespoke size exists.

**Comment by joshblack** [comment]
@vpicone if I'm querying for size information about an icon, why would size information not exist in `sizes`? Having to add in a distinct `glyph` field feels bizarre given that the purpose of the field is to communicate what sizes are available for this asset.

**Comment by vpicone** [comment]
@joshblack because querying for size is useless when they all exist at every size.

**Comment by joshblack** [comment]
@vpicone this may be the case now, but this was not always the case and may not be the case in the future. The field itself has lived through each of these decisions without us having to change up implementation details.

In addition, querying for size was simply the example here. Regardless of this exact function, it seems fair to suggest that it is valid and expected to enumerate all sizes in a `sizes` field.

**Comment by vpicone** [comment]
I’d say it’s equally valid to expect all of those sizes, then, to be a size. Not a string representing the idea of a bespoke size.

**Comment by joshblack** [comment]
> I’d say it’s equally valid to expect all of those sizes, then, to be a size. Not a string representing the idea of a bespoke size.

The size fields represents our decision around bounding boxes (16-32) and glyphs versus a pixel size. If we were typing it, this type would be:

```ts
type Size = 16 | 20 | 24 | 32 | 'glyph';
type Sizes = Array<Size>;
```

**Comment by vpicone** [comment]
I’d say that’s a pretty weak type though from both a type perspective (mixing strings and numbers) and a logical perspective (literal sizes vs a symbol of some size).  One of these things not like the other. Either way, not especially relevant to this PR, no sense holding it up.

**Comment by joshblack** [comment]
One usage of a type is to encode a domain object as a value in code such that a type checker can examine the flow of this value to attempt to prove no type errors can occur. If the domain object is encoded in this way, then it is unfair to critique the type as weak as it directly serves the purpose of representing the domain object in code.

**Review by vpicone — APPROVED** [review]

**Pull request merged**